src/matrix_observer.py

```diff
@@ -24,19 +31,37 @@ def __init__(self, charm: CharmBase, relation_name: str):
         super().__init__(charm, "matrix-observer")
         self._charm = charm
         self.relation_name = relation_name
+        self.matrix = MatrixAuthRequires(
+            self._charm,
+            relation_name=relation_name,
+        )
+        self.framework.observe(
+            self.matrix.on.matrix_auth_request_processed,
+            self._on_matrix_auth_request_processed,
+        )
 
-    def _get_relation_data(self) -> typing.Optional[DatasourceMatrix]:
-        """Get matrix data from relation.
+    def _on_matrix_auth_request_processed(self, _: Object) -> None:
+        """Handle the matrix auth request processed event."""
+        logger.info("Matrix auth request processed")
+        self._charm.reconcile()  # type: ignore
```

Comment:
If changing the type of the charm in the init function to IRCCharm is not desired, I think other options could be used instead of the ignore, like structural subtyping or an abstract class that is in another module.

Reply:
Fair point, I went with an abstract class to be consistent with synapse-operator.